kong/llm/init.lua

```diff
@@ -252,6 +361,190 @@ function _M.is_compatible(request, route_type)
+      end
+
+      local metadata
+
+      -- then parse each into the standard inference format
+      for i, event in ipairs(events) do
+        local event_t
+        local token_t
+
+        -- some LLMs do a final reply with token counts, and such
+        -- so we will stash them if supported
+        local formatted, err, this_metadata = self.driver.from_format(event, self.conf.model, "stream/" .. self.conf.route_type)
+        if err then
+          return internal_server_error(err)
+        end
+
+        metadata = this_metadata or metadata
+
+        -- handle event telemetry
+        if self.conf.logging.log_statistics then
+
+          if not ai_shared.streaming_has_token_counts[self.conf.model.provider] then
+            event_t = cjson.decode(formatted)
+
+            -- chat
+            token_t = event_t and
+                      event_t.choices and
+                      #event_t.choices > 0 and
+                      event_t.choices[1].delta and
+                      event_t.choices[1].delta.content
+
+                      or
+
+            -- completions
+                      event_t and
+                      event_t.choices and
+                      #event_t.choices > 0 and
+                      event_t.choices[1].text
+
+                      or ""
+
+            -- incredibly loose estimate based on https://help.openai.com/en/articles/4936856-what-are-tokens-and-how-to-count-them
+            -- but this is all we can do until OpenAI fixes this...
+            --
+            -- essentially, every 4 characters is a token, with minimum of 1 per event
+            telemetry_fake_table.usage.completion_tokens =
+                telemetry_fake_table.usage.completion_tokens + math.ceil(#string.gsub(token_t, "^%s*(.-)%s*$", "%1") / 4)
+
+          elseif metadata then
+            telemetry_fake_table.usage.completion_tokens = metadata.completion_tokens
+            telemetry_fake_table.usage.prompt_tokens = metadata.prompt_tokens
+          end
+
+        end
+
+        -- then stream to the client
+        if formatted then  -- only stream relevant frames back to the user
+          if self.conf.logging.log_payloads then
+            -- append the "choice" to the buffer, for logging later. this actually works!
+            if not event_t then
+              event_t = cjson.decode(formatted)
+            end
+            
+            if not token_t then
+              -- chat
+              token_t = event_t and
+                        event_t.choices and
+                        #event_t.choices > 0 and
+                        event_t.choices[1].delta and
+                        event_t.choices[1].delta.content
+
+                        or
+              
+              -- completions
+                        event_t and
+                        event_t.choices and
+                        #event_t.choices > 0 and
+                        event_t.choices[1].text
+
+                        or ""
```

Comment:
nitpicking: duplicate with line 460-474, shall we replace it with some helper function?

Reply:
Yes good idea

Fixed in 0cdecea6fd3e7ec378e2805f3807295078998f09